File: game/js/fx.js
// 이펙트: 파티클, 떠오르는 숫자, 화면 흔들림, 히트스톱, 슬로모션, 섬광
'use strict';
const FX = {
  parts: [], texts: [], sprites: [], bolts: [], rings: [],
  shakeAmt: 0, flashColor: '#fff', flashA: 0, hitstop: 0,
  slowT: 0, slowScale: 1, zoom: 1, zoomT: 0, zoomX: W / 2, zoomY: 500,
  clear() { this.parts = []; this.texts = []; this.sprites = []; this.bolts = []; this.rings = []; this.shakeAmt = 0; this.flashA = 0; this.hitstop = 0; this.slowT = 0; this.zoomT = 0; },

  shake(a) { this.shakeAmt = Math.max(this.shakeAmt, a); },
  flash(color, a = 0.6) { this.flashColor = color; this.flashA = Math.max(this.flashA, a); },
  stop(t) { this.hitstop = Math.max(this.hitstop, t); },
  slow(t, s = 0.25) { this.slowT = t; this.slowScale = s; },
  punch(x, y, t = 0.35) { this.zoomT = t; this.zoomX = x; this.zoomY = y; },
  timeScale() { return this.slowT > 0 ? this.slowScale : 1; },

  // 파티클 기본형
  p(o) {
    if (this.parts.length > 900) this.parts.shift();
    this.parts.push(Object.assign({ x: 0, y: 0, vx: 0, vy: 0, g: 0, drag: 0.9, life: 0.6, t: 0, size: 10, color: '#fff', add: true, shrink: true, rot: 0, vr: 0 }, o));
  },
  burst(x, y, color, n = 16, speed = 600, size = 12, life = 0.5) {
    for (let i = 0; i < n; i++) {
      const a = Math.random() * Math.PI * 2, s = rand(0.3, 1) * speed;
      this.p({ x, y, vx: Math.cos(a) * s, vy: Math.sin(a) * s, size: rand(0.5, 1) * size, color, life: rand(0.6, 1) * life, drag: 0.88 });
    }
  },
  sparks(x, y, color = '#ffd36b', n = 10, dir = 0, spread = Math.PI) {
    for (let i = 0; i < n; i++) {
      const a = dir + rand(-spread / 2, spread / 2), s = rand(400, 1100);
      this.p({ x, y, vx: Math.cos(a) * s, vy: Math.sin(a) * s, size: rand(3, 6), color, life: rand(0.2, 0.45), streak: true, g: 1400, drag: 0.93 });
    }
  },
  stars(x, y, n = 8, spread = 300, size = 40, ui = false) {
    for (let i = 0; i < n; i++) {
      const a = Math.random() * Math.PI * 2, s = rand(0.4, 1) * spread;
      this.p({ x, y, vx: Math.cos(a) * s, vy: Math.sin(a) * s - 200, g: 600, img: 'icon.star', size: rand(0.5, 1) * size, life: rand(0.5, 0.9), vr: rand(-8, 8), add: false, ui });
    }
  },
  smoke(x, y, n = 6, color = 'rgba(240,225,205,0.45)') {
    for (let i = 0; i < n; i++) this.p({ x: x + rand(-30, 30), y: y + rand(-10, 10), vx: rand(-120, 120), vy: rand(-160, -40), size: rand(14, 28), color, life: rand(0.4, 0.8), add: false, grow: true, drag: 0.92 });
  },
  ring(x, y, color, maxR = 250, life = 0.45, width = 18) { this.rings.push({ x, y, color, maxR, life, t: 0, width }); },
  bolt(x1, y1, x2, y2, color = '#bfe6ff', life = 0.25) {
    const pts = [];
    const n = 10;
    for (let i = 0; i <= n; i++) {
      const t = i / n;
      pts.push({ x: lerp(x1, x2, t) + (i && i < n ? rand(-40, 40) : 0), y: lerp(y1, y2, t) + (i && i < n ? rand(-20, 20) : 0) });
    }
    this.bolts.push({ pts, color, life, t: 0 });
  },
  // 이미지 스프라이트 이펙트(검기, 치유 오라, 합성 빛)
  sprite(key, x, y, w, h, o = {}) {
    this.sprites.push(Object.assign({ key, x, y, w, h, t: 0, life: 0.35, scale0: 0.6, scale1: 1.15, rot: 0, flip: false, add: false }, o));
  },
  number(x, y, value, o = {}) {
    if (this.texts.length > 80) this.texts.shift();
    this.texts.push(Object.assign({ x: x + rand(-25, 25), y, s: typeof value === 'number' ? fmt(value) : value, t: 0, life: 0.9, size: 44, color: '#fff', stroke: '#5a2a18', vy: -260 }, o));
  },

  update(dt) {
    this.shakeAmt = Math.max(0, this.shakeAmt - dt * 60);
    this.flashA = Math.max(0, this.flashA - dt * 2.5);
    if (this.slowT > 0) this.slowT -= dt;
    if (this.zoomT > 0) this.zoomT -= dt;
    for (const p of this.parts) {
      p.t += dt;
      p.vx *= Math.pow(p.drag, dt * 60); p.vy *= Math.pow(p.drag, dt * 60);
      p.vy += p.g * dt;
      if (p.target) {
        // 목표 지점으로 빨려 들어감
        const k = clamp((p.t - p.delay) / p.flyTime, 0, 1);
        if (p.t > p.delay) {
          if (p.sx === undefined) { p.sx = p.x; p.sy = p.y; }
          const e = k * k;
          p.x = lerp(p.sx, p.target.x, e) + Math.sin(k * Math.PI) * p.curve;
          p.y = lerp(p.sy, p.target.y, e);
          if (k >= 1 && !p.arrived) { p.arrived = true; p.t = p.life; p.onArrive && p.onArrive(); }
          continue;
        }
      }
      p.x += p.vx * dt; p.y += p.vy * dt; p.rot += p.vr * dt;
      if (p.floor !== undefined && p.y > p.floor) { p.y = p.floor; p.vy *= -0.45; p.vx *= 0.7; }
    }
    this.parts = this.parts.filter(p => p.t < p.life);
    for (const t of this.texts) { t.t += dt; t.y += t.vy * dt; t.vy *= Math.pow(0.9, dt * 60); }
    this.texts = this.texts.filter(t => t.t < t.life);
    for (const s of this.sprites) s.t += dt;
    this.sprites = this.sprites.filter(s => s.t < s.life);
    for (const b of this.bolts) b.t += dt;
    this.bolts = this.bolts.filter(b => b.t < b.life);
    for (const r of this.rings) r.t += dt;
    this.rings = this.rings.filter(r => r.t < r.life);
  },

  // 월드 좌표에 있는 이펙트 (흔들림 적용 레이어 안에서 호출)
  drawWorld() {
    for (const s of this.sprites) {
      const k = s.t / s.life, sc = lerp(s.scale0, s.scale1, easeOut(k));
      ctx.save();
      if (s.add) ctx.globalCompositeOperation = 'lighter';
      ctx.globalAlpha = k < 0.2 ? k / 0.2 : 1 - (k - 0.2) / 0.8;
      ctx.translate(s.x, s.y); ctx.rotate(s.rot); ctx.scale(sc * (s.flip ? -1 : 1), sc);
      img(s.key, -s.w / 2, -s.h / 2, s.w, s.h);
      ctx.restore();
    }
    for (const r of this.rings) {
      const k = r.t / r.life;
      ctx.save();
      ctx.globalCompositeOperation = 'lighter';
      ctx.globalAlpha = 1 - k;
      ctx.strokeStyle = r.color;
      ctx.lineWidth = r.width * (1 - k) + 2;
      ctx.beginPath(); ctx.arc(r.x, r.y, r.maxR * easeOut(k), 0, Math.PI * 2); ctx.stroke();
      ctx.restore();
    }
    for (const b of this.bolts) {
      ctx.save();
      ctx.globalCompositeOperation = 'lighter';
      ctx.globalAlpha = 1 - b.t / b.life;
      for (const [w, c] of [[22, b.color], [7, '#ffffff']]) {
        ctx.strokeStyle = c; ctx.lineWidth = w; ctx.lineJoin = 'round';
        ctx.shadowColor = b.color; ctx.shadowBlur = 30;
        ctx.beginPath();
        b.pts.forEach((p, i) => i ? ctx.lineTo(p.x, p.y) : ctx.moveTo(p.x, p.y));
        ctx.stroke();
      }
      ctx.restore();
    }
    this.drawParts(false);
  },
  drawParts(ui) {
    for (const p of this.parts) {
      if (!!p.ui !== ui) continue;
      const k = p.t / p.life;
      const a = p.fade === false ? 1 : k > 0.6 ? 1 - (k - 0.6) / 0.4 : 1;
      const size = p.grow ? p.size * (1 + k * 1.5) : p.shrink ? p.size * (1 - k * 0.7) : p.size;
      ctx.save();
      ctx.globalAlpha = a * (p.alpha || 1);
      if (p.img) {
        ctx.translate(p.x, p.y); ctx.rotate(p.rot);
        img(p.img, -size / 2, -size / 2, size, size);
      } else {
        if (p.add) ctx.globalCompositeOperation = 'lighter';
        ctx.fillStyle = p.color;
        if (p.streak) {
          ctx.strokeStyle = p.color; ctx.lineWidth = size; ctx.lineCap = 'round';
          ctx.beginPath(); ctx.moveTo(p.x, p.y); ctx.lineTo(p.x - p.vx * 0.04, p.y - p.vy * 0.04); ctx.stroke();
        } else {
          ctx.beginPath(); ctx.arc(p.x, p.y, Math.max(0.5, size), 0, Math.PI * 2); ctx.fill();
        }
      }
      ctx.restore();
    }
  },
  drawTexts() {
    for (const t of this.texts) {
      const k = t.t / t.life;
      const pop = t.t < 0.12 ? lerp(1.8, 1, t.t / 0.12) : 1;
      const a = k > 0.7 ? 1 - (k - 0.7) / 0.3 : 1;
      ctx.save();
      ctx.translate(t.x, t.y); ctx.scale(pop, pop);
      text(t.s, 0, 0, { size: t.size, color: t.color, stroke: t.stroke, alpha: a, weight: 900, shadow: t.glow, blur: 20 });
      ctx.restore();
    }
  },
  drawOverlay() {
    if (this.flashA > 0) {
      ctx.save();
      ctx.globalAlpha = this.flashA;
      ctx.fillStyle = this.flashColor;
      ctx.fillRect(0, 0, W, H);
      ctx.restore();
    }
  },
  applyCamera() {
    if (this.shakeAmt > 0) ctx.translate(rand(-1, 1) * this.shakeAmt, rand(-1, 1) * this.shakeAmt);
    if (this.zoomT > 0) {
      const z = 1 + 0.06 * Math.sin((this.zoomT / 0.35) * Math.PI);
      ctx.translate(this.zoomX, this.zoomY); ctx.scale(z, z); ctx.translate(-this.zoomX, -this.zoomY);
    }
  },
  // 골드·보석이 HUD로 빨려 들어감
  collect(x, y, key, target, n, onEach, ui = true) {
    for (let i = 0; i < n; i++) {
      const a = Math.random() * Math.PI * 2, s = rand(200, 650);
      this.p({ x, y, vx: Math.cos(a) * s, vy: Math.sin(a) * s - 300, g: 1500, drag: 0.9, img: key, size: rand(38, 52), life: 5, add: false, shrink: false, fade: false,
        target, delay: rand(0.35, 0.7), flyTime: rand(0.35, 0.55), curve: rand(-120, 120), onArrive: onEach, ui, vr: rand(-5, 5) });
    }
  },
};

File: native/game/src/main/kotlin/com/mergenyang/game/Fx.kt
package com.mergenyang.game

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.utils.Pool
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin

fun rnd(a: Float, b: Float) = MathUtils.random(a, b)
fun easeOut(t: Float) = 1f - (1f - t).pow(3)
fun easeOutBack(t: Float): Float {
    val c = 1.70158f
    return 1f + (c + 1f) * (t - 1f).pow(3) + c * (t - 1f).pow(2)
}

/** 파티클 1개. 오브젝트 풀로 재사용해 GC 끊김을 막는다. */
class Particle : Pool.Poolable {
    var x = 0f; var y = 0f; var vx = 0f; var vy = 0f
    var g = 0f; var drag = 0.9f; var life = 0.6f; var t = 0f
    var size = 10f; val color = Color(Color.WHITE)
    var additive = true; var shrink = true; var grow = false; var fade = true
    var streak = false; var ui = false
    var img: String? = null; var rot = 0f; var vr = 0f
    var floor = Float.NaN
    // 목표로 빨려 들어가는 파티클
    var target = false; var tx = 0f; var ty = 0f; var delay = 0f; var flyTime = 0.4f; var curve = 0f
    var sx = Float.NaN; var sy = 0f; var arrived = false
    var onArrive: (() -> Unit)? = null

    override fun reset() {
        x = 0f; y = 0f; vx = 0f; vy = 0f; g = 0f; drag = 0.9f; life = 0.6f; t = 0f
        size = 10f; color.set(Color.WHITE); additive = true; shrink = true; grow = false; fade = true
        streak = false; ui = false; img = null; rot = 0f; vr = 0f; floor = Float.NaN
        target = false; delay = 0f; flyTime = 0.4f; curve = 0f; sx = Float.NaN; arrived = false; onArrive = null
    }
}

class FloatText(
    var x: Float, var y: Float, val s: String, val size: Float, val color: Color,
    val life: Float, var vy: Float, val glow: Boolean,
) { var t = 0f }

class SpriteFx(
    val key: String, val x: Float, val y: Float, val w: Float, val h: Float,
    val life: Float, val scale0: Float, val scale1: Float, val rot: Float, val additive: Boolean,
) { var t = 0f }

class RingFx(val x: Float, val y: Float, val color: Color, val maxR: Float, val life: Float, val width: Float) { var t = 0f }

class BoltFx(val xs: FloatArray, val ys: FloatArray, val color: Color, val life: Float) { var t = 0f }

/** 화면 흔들림, 섬광, 히트스톱, 슬로모션, 파티클, 떠오르는 숫자 */
class Fx {
    private val pool = object : Pool<Particle>(256, 1200) {
        override fun newObject() = Particle()
    }
    val parts = ArrayList<Particle>(512)
    val texts = ArrayList<FloatText>()
    val sprites = ArrayList<SpriteFx>()
    val rings = ArrayList<RingFx>()
    val bolts = ArrayList<BoltFx>()

    var shakeAmt = 0f
    val flashColor = Color(Color.WHITE)
    var flashA = 0f
    var hitstop = 0f
    var slowT = 0f
    var slowScale = 1f
    var zoomT = 0f
    var zoomX = 540f
    var zoomY = 500f

    fun clear() {
        parts.forEach { pool.free(it) }
        parts.clear(); texts.clear(); sprites.clear(); rings.clear(); bolts.clear()
        shakeAmt = 0f; flashA = 0f; hitstop = 0f; slowT = 0f; zoomT = 0f
    }

    fun shake(a: Float) { shakeAmt = max(shakeAmt, a) }
    fun flash(c: Color, a: Float = 0.6f) { flashColor.set(c); flashA = max(flashA, a) }
    fun stop(t: Float) { hitstop = max(hitstop, t) }
    fun slow(t: Float, s: Float = 0.25f) { slowT = t; slowScale = s }
    fun punch(x: Float, y: Float, t: Float = 0.35f) { zoomT = t; zoomX = x; zoomY = y }
    fun timeScale() = if (slowT > 0f) slowScale else 1f

    fun p(init: Particle.() -> Unit): Particle {
        if (parts.size > 900) pool.free(parts.removeAt(0))
        val p = pool.obtain()
        p.init()
        parts += p
        return p
    }

    fun burst(x: Float, y: Float, c: Color, n: Int = 16, speed: Float = 600f, size: Float = 12f, life: Float = 0.5f) {
        repeat(n) {
            val a = rnd(0f, 2 * PI.toFloat())
            val s = rnd(0.3f, 1f) * speed
            p {
                this.x = x; this.y = y; vx = cos(a) * s; vy = sin(a) * s
                this.size = rnd(0.5f, 1f) * size; color.set(c); this.life = rnd(0.6f, 1f) * life; drag = 0.88f
            }
        }
    }

    fun sparks(x: Float, y: Float, c: Color, n: Int = 10, dir: Float = 0f, spread: Float = PI.toFloat()) {
        repeat(n) {
            val a = dir + rnd(-spread / 2, spread / 2)
            val s = rnd(400f, 1100f)
            p {
                this.x = x; this.y = y; vx = cos(a) * s; vy = sin(a) * s
                size = rnd(3f, 6f); color.set(c); life = rnd(0.2f, 0.45f); streak = true; g = 1400f; drag = 0.93f
            }
        }
    }

    fun stars(x: Float, y: Float, n: Int = 8, spread: Float = 300f, size: Float = 40f, ui: Boolean = false) {
        repeat(n) {
            val a = rnd(0f, 2 * PI.toFloat())
            val s = rnd(0.4f, 1f) * spread
            p {
                this.x = x; this.y = y; vx = cos(a) * s; vy = sin(a) * s - 200f; g = 600f
                img = "icons/star"; this.size = rnd(0.5f, 1f) * size; life = rnd(0.5f, 0.9f)
                vr = rnd(-400f, 400f); additive = false; this.ui = ui
            }
        }
    }

    fun smoke(x: Float, y: Float, n: Int = 6, c: Color = SMOKE) {
        repeat(n) {
            p {
                this.x = x + rnd(-30f, 30f); this.y = y + rnd(-10f, 10f); vx = rnd(-120f, 120f); vy = rnd(-160f, -40f)
                size = rnd(14f, 28f); color.set(c); life = rnd(0.4f, 0.8f); additive = false; grow = true; drag = 0.92f
            }
        }
    }

    fun ring(x: Float, y: Float, c: Color, maxR: Float = 250f, life: Float = 0.45f, width: Float = 18f) {
        rings += RingFx(x, y, Color(c), maxR, life, width)
    }

    fun bolt(x1: Float, y1: Float, x2: Float, y2: Float, c: Color, life: Float = 0.25f) {
        val n = 10
        val xs = FloatArray(n + 1)
        val ys = FloatArray(n + 1)
        for (i in 0..n) {
            val t = i / n.toFloat()
            val jitter = i in 1 until n
            xs[i] = x1 + (x2 - x1) * t + if (jitter) rnd(-40f, 40f) else 0f
            ys[i] = y1 + (y2 - y1) * t + if (jitter) rnd(-20f, 20f) else 0f
        }
        bolts += BoltFx(xs, ys, Color(c), life)
    }

    fun sprite(key: String, x: Float, y: Float, w: Float, h: Float, life: Float = 0.35f, scale0: Float = 0.6f, scale1: Float = 1.15f, rot: Float = 0f, additive: Boolean = false) {
        sprites += SpriteFx(key, x, y, w, h, life, scale0, scale1, rot, additive)
    }

    fun number(x: Float, y: Float, s: String, size: Float = 44f, color: Color = Color.WHITE, life: Float = 0.9f, vy: Float = -260f, glow: Boolean = false) {
        if (texts.size > 80) texts.removeAt(0)
        texts += FloatText(x + rnd(-25f, 25f), y, s, size, Color(color), life, vy, glow)
    }

    /** 골드·보석이 목표(HUD)로 빨려 들어감 */
    fun collect(x: Float, y: Float, key: String, tx: Float, ty: Float, n: Int, ui: Boolean, onEach: () -> Unit) {
        repeat(n) {
            val a = rnd(0f, 2 * PI.toFloat())
            val s = rnd(200f, 650f)
            p {
                this.x = x; this.y = y; vx = cos(a) * s; vy = sin(a) * s - 300f; g = 1500f; drag = 0.9f
                img = key; size = rnd(38f, 52f); life = 5f; additive = false; shrink = false; fade = false
                target = true; this.tx = tx; this.ty = ty; delay = rnd(0.35f, 0.7f); flyTime = rnd(0.35f, 0.55f)
                curve = rnd(-120f, 120f); onArrive = onEach; this.ui = ui; vr = rnd(-300f, 300f)
            }
        }
    }

    fun update(dt: Float) {
        shakeAmt = max(0f, shakeAmt - dt * 60f)
        flashA = max(0f, flashA - dt * 2.5f)
        if (slowT > 0f) slowT -= dt
        if (zoomT > 0f) zoomT -= dt
        val it = parts.iterator()
        while (it.hasNext()) {
            val p = it.next()
            p.t += dt
            val d = p.drag.pow(dt * 60f)
            p.vx *= d; p.vy *= d
            p.vy += p.g * dt
            if (p.target && p.t > p.delay) {
                if (p.sx.isNaN()) { p.sx = p.x; p.sy = p.y }
                val k = min(1f, (p.t - p.delay) / p.flyTime)
                val e = k * k
                p.x = p.sx + (p.tx - p.sx) * e + sin(k * PI.toFloat()) * p.curve
                p.y = p.sy + (p.ty - p.sy) * e
                if (k >= 1f && !p.arrived) {
                    p.arrived = true
                    p.onArrive?.invoke()
                    p.t = p.life
                }
            } else {
                p.x += p.vx * dt; p.y += p.vy * dt; p.rot += p.vr * dt
                if (!p.floor.isNaN() && p.y > p.floor) { p.y = p.floor; p.vy *= -0.45f; p.vx *= 0.7f }
            }
            if (p.t >= p.life) {
                it.remove()
                pool.free(p)
            }
        }
        texts.forEach { t -> t.t += dt; t.y += t.vy * dt; t.vy *= 0.9f.pow(dt * 60f) }
        texts.removeAll { t -> t.t >= t.life }
        sprites.forEach { s -> s.t += dt }
        sprites.removeAll { s -> s.t >= s.life }
        rings.forEach { r -> r.t += dt }
        rings.removeAll { r -> r.t >= r.life }
        bolts.forEach { b -> b.t += dt }
        bolts.removeAll { b -> b.t >= b.life }
    }

    fun drawWorld() {
        val b = Gfx.batch
        for (s in sprites) {
            val k = s.t / s.life
            val sc = s.scale0 + (s.scale1 - s.scale0) * easeOut(k)
            val a = if (k < 0.2f) k / 0.2f else 1f - (k - 0.2f) / 0.8f
            val r = Gfx.assets.region(s.key) ?: continue
            val w = s.w * sc
            val h = s.h * sc
            val draw = {
                b.setColor(1f, 1f, 1f, a)
                b.draw(r, s.x - w / 2, Gfx.H - s.y - h / 2, w / 2, h / 2, w, h, 1f, 1f, s.rot)
                b.setColor(Color.WHITE)
            }
            if (s.additive) Gfx.additive(draw) else draw()
        }
        Gfx.additive {
            for (r in rings) {
                val k = r.t / r.life
                Gfx.ring(r.x, r.y, r.maxR * easeOut(k) + 4f, r.color, (1f - k) * min(1f, r.width / 12f))
            }
            for (bo in bolts) {
                val a = 1f - bo.t / bo.life
                for (i in 0 until bo.xs.size - 1) {
                    Gfx.line(bo.xs[i], bo.ys[i], bo.xs[i + 1], bo.ys[i + 1], 26f, bo.color, a * 0.7f)
                    Gfx.line(bo.xs[i], bo.ys[i], bo.xs[i + 1], bo.ys[i + 1], 8f, Color.WHITE, a)
                }
            }
        }
        drawParts(false)
    }

    fun drawParts(ui: Boolean) {
        val b = Gfx.batch
        for (p in parts) {
            if (p.ui != ui) continue
            val k = p.t / p.life
            val a = if (!p.fade) 1f else if (k > 0.6f) 1f - (k - 0.6f) / 0.4f else 1f
            val size = when {
                p.grow -> p.size * (1f + k * 1.5f)
                p.shrink -> p.size * (1f - k * 0.7f)
                else -> p.size
            }
            val img = p.img
            if (img != null) {
                val r = Gfx.assets.region(img) ?: continue
                b.setColor(1f, 1f, 1f, a)
                b.draw(r, p.x - size / 2, Gfx.H - p.y - size / 2, size / 2, size / 2, size, size, 1f, 1f, p.rot)
                b.setColor(Color.WHITE)
            } else if (p.streak) {
                if (p.additive) Gfx.additive { Gfx.line(p.x, p.y, p.x - p.vx * 0.04f, p.y - p.vy * 0.04f, size, p.color, a) }
                else Gfx.line(p.x, p.y, p.x - p.vx * 0.04f, p.y - p.vy * 0.04f, size, p.color, a)
            } else {
                if (p.additive) Gfx.additive { Gfx.dot(p.x, p.y, size * 1.3f, p.color, a) }
                else Gfx.dot(p.x, p.y, size * 1.3f, p.color, a * p.color.a)
            }
        }
    }

    fun drawTexts() {
        for (t in texts) {
            val k = t.t / t.life
            val pop = if (t.t < 0.12f) 1.8f + (1f - 1.8f) * (t.t / 0.12f) else 1f
            val a = if (k > 0.7f) 1f - (k - 0.7f) / 0.3f else 1f
            if (t.glow) Gfx.additive { Gfx.glow(t.x, t.y, t.size * 1.3f, ORANGE, a * 0.5f) }
            Gfx.text(t.s, t.x, t.y, t.size, t.color, outline = true, alpha = a, scale = pop)
        }
    }

    fun drawOverlay() {
        if (flashA > 0f) Gfx.fullRect(flashColor, flashA)
    }

    /** 흔들림·줌을 카메라 오프셋으로 */
    fun cameraOffset(out: FloatArray) {
        out[0] = if (shakeAmt > 0) rnd(-1f, 1f) * shakeAmt else 0f
        out[1] = if (shakeAmt > 0) rnd(-1f, 1f) * shakeAmt else 0f
        out[2] = if (zoomT > 0f) 1f + 0.06f * sin((zoomT / 0.35f) * PI.toFloat()) else 1f
    }

    companion object {
        val SMOKE = Color(0.94f, 0.88f, 0.8f, 0.45f)
        val ORANGE: Color = Color.valueOf("ff8a00ff")
    }
}
